src/realm/status.cpp

```diff
@@ -0,0 +1,80 @@
+/*************************************************************************
+ *
+ * Copyright 2021 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#include "realm/status.hpp"
+
+#include "realm/util/demangle.hpp"
+
+#include <iostream>
+
+namespace realm {
+
+Status::Status(ErrorCodes::Error code, StringData reason)
+    : m_error(ErrorInfo::create(code, reason))
+{
+    // OK status should be created by calling Status::OK() - which is a special case that doesn't allocate
+    // anything.
+    REALM_ASSERT(code != ErrorCodes::OK);
+}
+
+Status::Status(ErrorCodes::Error code, const std::string& reason)
+    : Status(code, StringData(reason))
+{
+}
+
+Status::Status(ErrorCodes::Error code, const char* reason)
+    : Status(code, StringData(reason, std::char_traits<char>::length(reason)))
+{
+}
+
+Status::ErrorInfo::ErrorInfo(ErrorCodes::Error code, StringData reason)
+    : m_refs(0)
+    , m_code(code)
+    , m_reason(reason)
+{
+}
+
+util::bind_ptr<Status::ErrorInfo> Status::ErrorInfo::create(ErrorCodes::Error code, StringData reason)
+{
+    return util::bind_ptr<Status::ErrorInfo>(new ErrorInfo(code, reason));
+}
+
+std::ostream& operator<<(std::ostream& out, const Status& val)
+{
+    out << val.code_string() << ": " << val.reason();
+    return out;
+}
+
+Status exception_to_status() noexcept
+{
+    try {
+        throw;
+    }
+    catch (const ExceptionForStatus& e) {
+        return e.to_status();
+    }
+    catch (const std::exception& e) {
+        return Status(ErrorCodes::UnknownError,
+                      util::format("Caught std::exception of type %1: %2", util::get_type_name(e), e.what()));
+    }
+    catch (...) {
+        REALM_UNREACHABLE();
```

Comment:
This makes me slightly uneasy. Yes it is good practice to only throw exceptions but I don't know that this is the right place to hard stop if something else is thrown. What do you think about making it more permissive?

Reply:
I think, in practice, this won't be an issue. In the [mongo codebase](https://github.com/mongodb/mongo/blob/8fdfe55ebd9bd4a9cd982e8eeb6052e6aedc4ccd/src/mongo/util/assert_util.cpp#L355) this function also handles boost::exception's, but we don't use boost in the realm code base. I think if folks are throwing exceptions that don't inherit from std::exception, then that's a bug and we should treat it as an invariant failure. Like, I don't know what we would even do with the exception here besides crashing.

Reply:
From a cursory search, there are a few existing places that we'd need to change to std::exceptions (ex [here](https://github.com/realm/realm-core/blob/master/src/realm/alloc_slab.cpp#L864) and [here](https://github.com/realm/realm-core/blob/master/src/realm/object-store/sync/generic_network_transport.hpp#L167)). While I agree that these should be changed, (and you will probably encounter others with the upcoming error handling work) they are not necessarily existing bugs and this could make them crashes when this pattern is adopted more widely. Presumably SDK's have some more graceful way of handling non-descript exceptions that reach the top level than aborting, or at least we should give them the option to have more control over that. One non-fatal option to handle these would be to create an `ErrorCodes::UnknownError` status. On the other hand, I can see value in enforcing an invariant around exceptions so this is not a blocking comment, I just want to make sure this is intentional and not an accidental behavioural change.